observation_portal/common/test_state_changes.py

```diff
@@ -394,12 +396,55 @@ def test_configuration_statuses_not_complete_or_failed_use_initial(self):
             dmixer.blend(Summary, configuration_status=cs, time_completed=0)
         initial_state = 'INITIAL'
         request_state = get_request_state_from_configuration_statuses(
-            initial_state, 100, observation.configuration_statuses.all()
+            initial_state, self.request, observation.configuration_statuses.all()
         )
         self.assertEqual(request_state, initial_state)
         # requestgroup state is not completed, no email sent
         self.assertEqual(len(mail.outbox), 0)
 
+    @override_settings(REQUEST_DETAIL_URL='test_url/{request_id}', MAX_RETRIES_PER_REQUEST=2)
+    def test_max_observations_failed_request_retry_limit(self):
+        # First observation will be marked as failed
+        observation = dmixer.blend(
+            Observation, request=self.request, start=self.now - timedelta(minutes=30),
+            end=self.now + timedelta(minutes=20), state='PENDING'
+        )
+        for configuration in self.request.configurations.all():
+            cs = dmixer.blend(
+                ConfigurationStatus, observation=observation, configuration=configuration, state='FAILED'
+            )
+            dmixer.blend(Summary, configuration_status=cs, time_completed=0)
+        observation.refresh_from_db()
+        self.assertEqual(observation.state, 'FAILED')
+        initial_state = 'INITIAL'
+        # But request is still in its initial state
+        request_state = get_request_state_from_configuration_statuses(
+            initial_state, self.request, observation.configuration_statuses.all()
+        )
+        self.assertEqual(request_state, initial_state)
+        # Second observation failing will trigger RETRY_LIMIT of 2
+        observation2 = dmixer.blend(
+            Observation, request=self.request, start=self.now - timedelta(minutes=30),
+            end=self.now + timedelta(minutes=20), state='PENDING'
+        )
+        for configuration in self.request.configurations.all():
+            cs = dmixer.blend(
+                ConfigurationStatus, observation=observation2, configuration=configuration, state='FAILED'
+            )
+            dmixer.blend(Summary, configuration_status=cs, time_completed=0)
+        observation2.refresh_from_db()
+        self.assertEqual(observation2.state, 'FAILED')
+        # so request should enter the RETRY_LIMIT state since it has failed "max retries" times
+        request_state = get_request_state_from_configuration_statuses(
+            initial_state, self.request, observation2.configuration_statuses.all()
+        )
+        self.assertEqual(request_state, 'RETRY_LIMIT')
+        # Test that an email is sent out for the Request
+        self.assertEqual(len(mail.outbox), 1)
```

Comment:
You should wait for the dramatiq worker to finish before checking what's in the mailbox: https://github.com/Bogdanp/django_dramatiq#using-unittest